### PR TITLE
Fix #1086-Added sign check on Constant f64 PartialEq implementation

### LINKED ISSUE #1086: False positive on `if pos { 0.0 } else { -0.0 }`
```
serde-rs/json/json/src/de.rs:438:47: 438:55 error: this `if` has identical blocks
serde-rs/json/json/src/de.rs:438         visitor.visit_f64(if pos { 0.0 } else { -0.0 })
                                                                               ^~~~~~~~
```

These are not identical.

### PR DESCRIPTION
**Pull request by cramertj**
Fix #1086

### CODE REVIEW

**Comment by oli-obk** [comment]
as @mcarton said, we should probably check for bit-equality (by transmuting to an int before the comparison), to make sure that any `NaN`s we reach are equal. Note that rustc compares constant floats by transmuting them to an int.

**Comment by cramertj** [comment]
@oli-obk Ah, missed the `NaN` bit. I thought it was just sign issue, and this seemed like a cleaner fix. I'll change it.

**Comment by oli-obk** [comment]
can you add tests for NaN, too? Here's an example on how to get two different NaNs:

https://is.gd/sWEtCH

division by zero and negative sqrt yield different NaN.

File: clippy_lints/src/consts.rs [hunk 1]
```diff
@@ -92,7 +92,10 @@ impl PartialEq for Constant {
                 // we want `Fw32 == FwAny` and `FwAny == Fw64`, by transitivity we must have
                 // `Fw32 == Fw64` so don’t compare them
                 match (ls.parse::<f64>(), rs.parse::<f64>()) {
-                    (Ok(l), Ok(r)) => l.eq(&r),
+                    // mem::transmute is required to catch non-matching 0.0, -0.0, and NaNs
+                    (Ok(l), Ok(r)) => unsafe {
+                        mem::transmute::<f64, u64>(l) == mem::transmute::<f64, u64>(r)
+                    },
```
Comment (mcarton) [inline clippy_lints/src/consts.rs]: You *must* change `PartialOrd` accordingly (`Hash` already does this).

File: tests/compile-fail/copies.rs [hunk 2]
```diff
@@ -249,6 +249,21 @@ fn if_same_then_else() -> Result<&'static str, ()> {
         -0.0
     };
 
+    // Different NaNs
+    let _ = if true {
+        1.0 / 0.0
+    } else {
+        (-5f32).sqrt()
```
Comment (mcarton) [inline tests/compile-fail/copies.rs]: `(-5f32).sqrt()` is not a constant expression and `1.0/0.0` is infinity, not NaN. So it's not affected by that bug anyway.
`0.0 / 0.0` vs. `std::f32::NAN` would have been more relevant.

File: clippy_lints/src/consts.rs (hunk 1 above)
Comment (cramertj) [inline clippy_lints/src/consts.rs]: Good catch!

File: tests/compile-fail/copies.rs (hunk 2 above)
Comment (cramertj) [inline tests/compile-fail/copies.rs]: Oh, I see. I was getting the values from the link @oli-obk posted and didn't question that they were all NaN. I will update it.

**Comment by oli-obk** [comment]
thanks!

**Pull request merged**